File: packages/coc-clang-format-style-options/scripts/parse.py
from typing import Dict, List
import re

# https://github.com/llvm/llvm-project/blob/25f753c51e7b17bfca08155c1d777c5667110970/clang/docs/ClangFormatStyleOptions.rst#configurable-format-style-options

# import pdb
import os

options = {}

lines = open(
    f"{os.path.dirname(__file__)}/ClangFormatStyleOptions.rst", "r"
).readlines()
pointer = 0
# pdb.set_trace()

def main():
    global pointer
    while True:
        line0 = lines[pointer]
        pointer += 1
        match0 = re.match(r"\*\*(\w+)\*\*\s\(``(.+?)``\)", line0)
        if match0:
            name = match0[1]
            type = match0[2]
            options[name] = {
                "type": type,
                "description": [],
            }
            has_code_block2 = False
            while True:
                line1 = lines[pointer]
                pointer += 1
                match1 = re.match(r"^\s+Possible values:$", line1)
                if match1:
                    options[name]["description"].append("    Possible values:\n\n")
                    options[name]["enum"] = {}
                    while True:
                        line2 = lines[pointer]
                        pointer += 1
                        match2 = re.match(r"^\s+\*\s``(.*?)``", line2)
                        if match2:
                            val_name = match2[1]
                            options[name]["enum"][val_name] = []
                            has_code_block = False
                            while True:
                                line3 = lines[pointer]
                                pointer += 1
                                # print(line3)
                                if re.match(r"^\s+\*\s``(.*?)``", line3):
                                    pointer -= 1
                                    if has_code_block:
                                        options[name]["enum"][val_name].append('```\n\n')
                                    break
                                if re.match(r"\*\*(\w+)\*\*\s\(``(.+?)``\)", line3):
                                    # print(line3)
                                    pointer -= 1
                                    if has_code_block:
                                        options[name]["enum"][val_name].append('```\n\n')
                                    break
                                if re.match(r"^\s+\.\. code-block:: .*", line3):
                                    line3 = line3.replace('c++', 'cpp')
                                    line3 = re.sub(r"^\s+\.\. code-block:: (\w+?)", r'```\1', line3)
                                    has_code_block = True
                                options[name]["enum"][val_name].append(line3)
                        if re.match(r"\*\*(\w+)\*\*\s\(``(.+?)``\)", line2):
                            pointer -= 1
                            break
                elif not re.match(r"\*\*(\w+)\*\*\s\(``(.+?)``\)", line1):
                    if re.match(r"^\s+\.\. code-block:: .*", line1):
                        line1 = line1.replace('c++', 'cpp')
                        line1 = re.sub(r"^\s+\.\. code-block:: (\w+?)", r'```\1', line1)
                        has_code_block2 = True
                    options[name]["description"].append(line1)
                if re.match(r"\*\*(\w+)\*\*\s\(``(.+?)``\)", line1):
                    if has_code_block2:
                        options[name]["description"].append('```')
                    pointer -= 1
                    break

try:
    main()
except Exception as e:
    print(f"ERROR")
    import json

    for name in options:
        option = options[name]
        description: List = option["description"]
        options[name]["description"] = "".join(description)
        if "enum" in option:
            for val_name in option["enum"]:
                value = option["enum"][val_name]
                print(value)
                options[name]["enum"][val_name] = "".join(value)

    with open("ClangFormatStyleOptions.json", "w") as f:
        json.dump(options, f)

File: packages/coc-clang-format-style-options/scripts/parse.ts
import fs from 'fs'
import { exit } from 'process'

const fsp = fs.promises

//

interface Values {
  name: string
  description: string
}

interface Option {
  name: string
  type: string
  description?: string
  values?: Values
}

const options: Option[] = []

let lines: string[] = [] //
let index = 0 // global file line pointer
const re_option_name_type = /\*\*(\w+)\*\* \(``(\w+)``\)/
const re_option_value = /^\s+Possible values:$/

async function do_options() {}

async function do_option(): Promise<Option> {
  const option = {} as Option

  const [name, type] = await do_option_name_type()
  option.name = name
  option.type = type

  const description = await do_option_description()
  if (description?.length > 0) option.description = description

  return option
}

async function do_option_name_type(): Promise<string[]> {
  const line = lines[index++]
  const group = line.match(re_option_name_type)
  const [_, name, type] = group
  return [name, type]
}

async function do_option_description(): Promise<string> {
  const description = []

  while (true) {
    const line = lines[index++]
    if (re_option_value.test(line)) {
      break
    } else if (re_option_name_type.test(line)) {
      break
    }
    description.push(line)
  }

  return description.join()
}

async function do_values(): Promise<Values> {
  const values = {} as Values
}

async function do_value_name(): Promise<string> {}

async function do_value_description(): Promise<string> {}

async function main() {
  const f = await fsp.readFile('./ClangFormatStyleOptions.rst', 'utf-8')
  lines = f.split('\n')
  for (const line of lines) {
    if (re_option_name_type.test(line)) {
      do_option()
      return
    }
    index++
  }
}

main().catch(console.error)
